feat(client): Added tweet sharing ability for a certification.

DZuz14 opened the pull request:
Checklist:
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #38194

I have added a Twitter icon that allows a user to tweet their certification. I decided to break down the initial `ShowCertification` client route by extracting the code to display an individual certification into a separate component named `Certificate`.

You'll see in the updated `ShowCertifications` file that I attach the props to the `Certificate` component via:
- `<Certificate {...cert} certName={certName} />`

The reason why I added the certName prop explicitly is because in the `ShowCertifications` file we have the following props:

```
const {
  cert,
  fetchState,
  validCertName,
  createFlashMessage,
  certName
} = this.props;
```

According to the propTypes, `cert` should have a certName property on it, but it was not being utilized. I figured there must be something I am not privy to, so I decided to make sure that I used all props in the same manner they were implemented in the prior version.

P.S. _I changed the wording a tiny bit, because perhaps a Facebook Icon could be added as well?_

Here are two screenshots of the twitter button for desktop and mobile.

![tweet-cert-desktop](https://user-images.githubusercontent.com/19492185/77425025-76020a80-6da8-11ea-85fc-d714ede9e3bd.jpg)

![tweet-cert-mobile](https://user-images.githubusercontent.com/19492185/77425028-769aa100-6da8-11ea-9bd2-c6e83579cb55.jpg)

DZuz14 committed:
feat: Added tweet sharing ability for a certification.

DZuz14 commented on client/src/components/profile/components/Certificate.js:
```diff
@@ -0,0 +1,104 @@
+  username: PropTypes.string
+};
+
+const CERT_URL = ' https://www.freecodecamp.org/certification';
```
That whitespace in CERT_URL is a mistake.

ahmadabdolsaheb commented:
Thanks @DZuz14 for looking into this. The functionality is there, and the UI needs a little bit of rearrangement. 

You could pull https://www.freecodecamp.org part of the cert url from env.json, so the functionality could be  tested when the build is deployed to .dev.

The message and the button could be displayed in a banner separated from the certificate itself. (a div similar to the flash message, but not a flash massage that is displayed at the top of the certificate.)

The button (banner) should only be displayed to the user when they are signed in and viewing their own certificate. for that logic please refer to the donation section in showCertification.

The banner could be displayed between donation section and certificate.

Also for some reason, the twitter icon was out of the button when I pulled and tested this feature locally.
![image](https://user-images.githubusercontent.com/4591597/77742327-617f7580-7027-11ea-9f07-de11bd80041a.png)

raisedadead commented:
I think we should do regular buttons, like regular [Share on Twitter] following our brand design philosophy.

Also, the donation section and the certificate it self need not be full-width IMHO. They should follow the usual page layout guidelines.

What do you think?

DZuz14 commented:
@ahmadabdolsaheb I read your comments and all that seems good! That was certainly interesting how that icon looked as well. What browser do you use? Just curious.

@raisedadead Do you guys have those icons laying around? I can look for them in the repo when I get a chance. The certificate following the normal guidelines makes sense, I can fix that.

DZuz14 commented:
I also have some questions about whether you guys might have some type of seed file so I can use an account on local development that says I have earned all of the certifications? I was kind of hacking things together by importing the Certificate component into a random page and just assigning it a bunch of defaultProps so I could work on it.

DZuz14 commented:
I won't be around for the rest of the day, but I was wondering if you guys were okay if I refactored the ShowCertification section into something like this as well?

Extracted the redirects logic in render into a handleRedirects method, created another component called DonationSection, and created a helper method to check if the user is the currently logged in one.

```
/* eslint-disable react/jsx-sort-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import DonationSection from '../components/profile/components/DonationSection';
import Certificate from '../components/profile/components/Certificate';

import {
  showCertSelector,
  showCertFetchStateSelector,
  showCert,
  userFetchStateSelector,
  usernameSelector,
  isDonatingSelector,
  executeGA
} from '../redux';
import validCertNames from '../../utils/validCertNames';
import { createFlashMessage } from '../components/Flash/redux';
import standardErrorMessage from '../utils/standardErrorMessage';
import reallyWeirdErrorMessage from '../utils/reallyWeirdErrorMessage';

import RedirectHome from '../components/RedirectHome';
import { Loader } from '../components/helpers';
import DonationSection from '../components/profile/components/DonationSection';

const propTypes = {
  cert: PropTypes.shape({
    username: PropTypes.string,
    name: PropTypes.string,
    certName: PropTypes.string,
    certTitle: PropTypes.string,
    completionTime: PropTypes.number,
    date: PropTypes.string
  }),
  certDashedName: PropTypes.string,
  certName: PropTypes.string,
  createFlashMessage: PropTypes.func.isRequired,
  executeGA: PropTypes.func,
  fetchState: PropTypes.shape({
    pending: PropTypes.bool,
    complete: PropTypes.bool,
    errored: PropTypes.bool
  }),
  isDonating: PropTypes.bool,
  issueDate: PropTypes.string,
  showCert: PropTypes.func.isRequired,
  signedInUserName: PropTypes.string,
  userFetchState: PropTypes.shape({
    complete: PropTypes.bool
  }),
  userFullName: PropTypes.string,
  username: PropTypes.string,
  validCertName: PropTypes.bool
};

const mapStateToProps = (state, { certName }) => {
  const validCertName = validCertNames.some(name => name === certName);
  return createSelector(
    showCertSelector,
    showCertFetchStateSelector,
    usernameSelector,
    userFetchStateSelector,
    isDonatingSelector,
    (cert, fetchState, signedInUserName, userFetchState, isDonating) => ({
      cert,
      fetchState,
      validCertName,
      signedInUserName,
      userFetchState,
      isDonating
    })
  );
};

const mapDispatchToProps = dispatch =>
  bindActionCreators({ createFlashMessage, showCert, executeGA }, dispatch);

class ShowCertification extends Component {
  constructor(...args) {
    super(...args);

    this.state = {
      isDonationSubmitted: false,
      isDonationDisplayed: false,
      isDonationClosed: false
    };

    this.hideDonationSection = this.hideDonationSection.bind(this);
    this.handleProcessing = this.handleProcessing.bind(this);
  }

  componentDidMount() {
    const { username, certName, validCertName, showCert } = this.props;
    if (validCertName) {
      return showCert({ username, certName });
    }
    return null;
  }

  shouldComponentUpdate(nextProps) {
    const { isDonating, executeGA } = nextProps;
    const { isDonationDisplayed } = this.state;

    if (!isDonationDisplayed && !isDonating && this.isLoggedInUser(nextProps)) {
      this.setState({
        isDonationDisplayed: true
      });

      executeGA({
        type: 'event',
        data: {
          category: 'Donation',
          action: 'Displayed Certificate Donation',
          nonInteraction: true
        }
      });
    }
    return true;
  }

  hideDonationSection() {
    this.setState({ isDonationDisplayed: false, isDonationClosed: true });
  }

  handleProcessing(duration, amount, action = 'stripe form submission') {
    this.props.executeGA({
      type: 'event',
      data: {
        category: 'donation',
        action: `certificate ${action}`,
        label: duration,
        value: amount
      }
    });
    this.setState({ isDonationSubmitted: true });
  }

  // props can be current props or nextProps
  isLoggedInUser(props) {
    const {
      userFetchState: { complete: userComplete },
      signedInUserName,
      cert: { username = '' }
    } = props;

    return userComplete && signedInUserName && signedInUserName === username;
  }

  handleRedirects() {
    const { fetchState, validCertName, createFlashMessage } = this.props;

    if (!validCertName) {
      createFlashMessage(standardErrorMessage);
      return <RedirectHome />;
    }

    const { pending, complete, errored } = fetchState;

    if (!pending && errored) {
      createFlashMessage(standardErrorMessage);
      return <RedirectHome />;
    }

    if (!pending && !complete && !errored) {
      createFlashMessage(reallyWeirdErrorMessage);
      return <RedirectHome />;
    }
  }

  render() {
    const { cert, certName } = this.props;

    const {
      isDonationSubmitted,
      isDonationDisplayed,
      isDonationClosed
    } = this.state;

    {
      this.handleRedirects();
    }

    if (pending) {
      return <Loader fullScreen={true} />;
    }

    return (
      <div className='certificate-outer-wrapper'>
        {isDonationDisplayed && !isDonationClosed && (
          <DonationSection
            submitted={isDonationSubmitted}
            handleProcessing={this.handleProcessing}
            close={this.hideDonationSection}
          />
        )}

        {this.isLoggedInUser(this.props) && (
          <Certificate {...cert} certName={certName} />
        )}
      </div>
    );
  }
}

ShowCertification.displayName = 'ShowCertification';
ShowCertification.propTypes = propTypes;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ShowCertification);

```

ahmadabdolsaheb commented:
I am not sure what you are referring to with the blue header, but the button and its wrapper (banner) should only be visible to a user if she is signed in and viewing her certificate.

Creating a component for donation and separating the logic is a good idea.

The certificate and the donation section could be adjusted to follow along the grid system as long as they don't squeezed in smaller screen.

Also if you decide to go down that route, please have individual commits for each feature/or fix.

ahmadabdolsaheb commented:
@DZuz14, is there any progress on this feature?

raisedadead commented:
Please re-open a fresh request when ready. Closing just to cleanup our backlog. Thanks.
